### ISSUE TITLE
Should not detrend go before low and high-pass filtering too for confounders in signal.clean?

### ISSUE THREAD

**Issue by jrasero** [nilearn/nilearn#2730]
Hi,

I don't know whether this is an issue or not, but I am trying to see if the deconfounding provided by the function _**signal.clean**_ works effectively or not on my data and I think it is not behaving as I expect, in particular when I combine it with a high-pass filtering. I know there's been a lot of issues around all these, with a recent pull-request #2609, which however does not seem to resolve my (apparent) issue.

Since **_signal.clean_** also applies the high-pass filtering to the confounds in order to keep this filter orthogonal, in order to see if the confounds removal works, I need to check that the correlation between the cleaned signal and the detrended and high-pass filtered confounders is indeed zero. 

Here is what I've done and where the issue arises. First, just with detrending and then confounds removal:

```
import numpy as np
from nilearn.signal import clean
import matplotlib.pylab as plt

corrs_x_c = []
for ii in range(1000):
    c = np.random.randn(100)[:,None]
    X = c + np.random.randn(100)[:,None]

    # Detrend + deconfounding
    X_clean = clean(X, detrend = True, high_pass = None, standardize_confounds=True, standardize=False, confounds=c)
    c_clean = clean(c, detrend = True, high_pass = None, standardize=False)
    corrs_x_c.append(np.corrcoef(X_clean.T, c_clean.T)[0,1])
plt.hist(corrs_x_c)
plt.title("Correlation between X_clean and c_clean")
```

![image](https://user-images.githubusercontent.com/25054369/110956473-a96ed980-8318-11eb-8107-84b90ae89fd1.png)

So here, the behaviour is the expected one. 

The problem is when I insert a high-pass filtering (and I guess with a low-pass will happen too). Following the same example:

```
import numpy as np
from nilearn.signal import clean
import matplotlib.pylab as plt

corrs_x_c = []
for ii in range(1000):
    c = np.random.randn(100)[:,None]
    X = c + np.random.randn(100)[:,None]

    # Detrend + High-pass filtering + deconfounding
    X_clean = clean(X, detrend=True, high_pass = 0.01, standardize_confounds=True, standardize=False, confounds=c)
    c_clean = clean(c, detrend=True, high_pass = 0.01, standardize=False)
    corrs_x_c.append(np.corrcoef(X_clean.T, c_clean.T)[0,1])
plt.hist(corrs_x_c)
plt.title("Correlation between X_clean and c_clean")
```

![image](https://user-images.githubusercontent.com/25054369/110958370-a2e16180-831a-11eb-9c0d-d59cc8448bee.png)

So as you can see, clean hasn't removed the confounders effectively from the signal.

I think the problem comes from the low and high-pass filtering, which, in the current code of the **_clean_** function, is performed before detrending (around line 600 in signal.py):

```
# Remove confounds
    if confounds is not None:
        confounds = _ensure_float(confounds)
        # Apply low- and high-pass filters to keep filters orthogonal
        # (according to Lindquist et al. (2018))
        if low_pass is not None or high_pass is not None:

            confounds = butterworth(confounds, sampling_rate=1. / t_r,
                                    low_pass=low_pass, high_pass=high_pass)

        confounds = _standardize(confounds, standardize=standardize_confounds,
                                 detrend=detrend)
```

However, for the signal, the order is first detrending and then high-pass filtering. Shouldn't the confounders follow the same order too?

In fact, modifying locally the original  _**clean**_ function in signal.py to the following:

```

# Remove confounds
if confounds is not None:
    confounds = _ensure_float(confounds)
    
    # Orthogonalise confounders to detreding first
    if detrend:
        confounds = _standardize(confounds, standardize=False,
                                    detrend=detrend)

    # Apply low- and high-pass filters to keep filters orthogonal
    # (according to Lindquist et al. (2018))    
    if low_pass is not None or high_pass is not None:

        confounds = butterworth(confounds, sampling_rate=1. / t_r,
                                low_pass=low_pass, high_pass=high_pass)

    confounds = _standardize(confounds, standardize=standardize_confounds,
                                detrend=False)
```

It solves my issue above. 

Does this make sense?

**Comment by NicolasGensollen** [comment]
Thanks for opening this @jrasero ! :+1: 
This makes sense and I was able to reproduce the bug.
I think having the confounders detrended before filtering, like you explained, works. 
Changing the code of `clean` to:

```python
# Remove confounds
if confounds is not None:
    confounds = _ensure_float(confounds)
    confounds = _standardize(confounds, standardize=standardize_confounds,
                             detrend=detrend)

    if not standardize_confounds:
        # Improve numerical stability by controlling the range of
        # confounds. We don't rely on _standardize as it removes any
        # constant contribution to confounds.
        confound_max = np.max(np.abs(confounds), axis=0)
        confound_max[confound_max == 0] = 1
        confounds /= confound_max

    # Apply low- and high-pass filters to keep filters orthogonal
    # (according to Lindquist et al. (2018))
    if low_pass is not None or high_pass is not None:
        confounds = butterworth(confounds, sampling_rate=1. / t_r,
                                low_pass=low_pass, high_pass=high_pass)
```

outputs the following on your example:

![signal](https://user-images.githubusercontent.com/2639645/110971032-a72d6f00-835a-11eb-9dd8-c3bb1a8d198e.png)

Do you want to open a PR to fix this?

**Comment by jrasero** [comment]
Great! 

The problem with your code is that correlations are still not as small as I expect. 

In fact, I modified the code of `clean` in a different way:

```
# Remove confounds
if confounds is not None:
    confounds = _ensure_float(confounds)
    
    # Detrend first to keep also this orthogonal
    if detrend:
        confounds = _standardize(confounds, standardize=False,
                                 detrend=detrend)
    
    # Apply low- and high-pass filters to keep filters orthogonal
    # (according to Lindquist et al. (2018))
    if low_pass is not None or high_pass is not None:

        confounds = butterworth(confounds, sampling_rate=1. / t_r,
                                low_pass=low_pass, high_pass=high_pass)

    confounds = _standardize(confounds, standardize=standardize_confounds,
                             detrend=False)

    if not standardize_confounds:
        # Improve numerical stability by controlling the range of
        # confounds. We don't rely on _standardize as it removes any
        # constant contribution to confounds.
        confound_max = np.max(np.abs(confounds), axis=0)
        confound_max[confound_max == 0] = 1
        confounds /= confound_max

```
With this, the output gives what I was expecting:

![image](https://user-images.githubusercontent.com/25054369/110972302-d8417b80-8329-11eb-9c40-c9bbf3d853cf.png)

The difference is that confounds standardization takes place after detrending and low and high-pass filtering. In your example, you low and high-pass filter the already standardized confounders, which presumably, after filtering they will no longer be standardized. Does it make sense?

**Comment by NicolasGensollen** [comment]
Thanks for the great explanation @jrasero !
Indeed, I didn't read your first code snippet with enough care, and only swapped detrending and filtering in mine...

**Comment by NicolasGensollen** [comment]
Fixed in #2732
